rfcs/20210119-determinism.md

```diff
@@ -0,0 +1,82 @@
+
+### Grappler changes
+Grappler graph optimizations may add nondeterministic behavior. In particular some optimizations will time out if they take too long to run. When determinism is enabled, these timeouts will be disabled.
+
+### Random ops
+Legacy random ops, such as `tf.random.normal`, are not deterministic if no seed is set, and so such ops will raise an error when determinism is enabled. To fix, the user should set a global seed with `tf.random.set_seed`. Since most models use legacy random ops, in practice users must call `tf.random.set_seed` when enabling deterministic behavior. Alternatively, users can pass a seed to every individual random operation, but doing so is more inconvenient.
+
+Certain random ops, such as `tf.image.sample_distorted_bounding_box` and `tf.nn.fractional_max_pool`, ignore the global seed if a seed is not explicitly passed. For such ops, setting the global seed is not enough to avoid the error, so users must pass a seed directly to the op.
+
+As for TensorFlow 2 random number generation, `tf.random.Generator.from_non_deterministic_state` will raise an error if called when determinism is enabled. In such cases, users should check if determinism is enabled and if so, use a different generator from a deterministic source. `tf.random.get_global_generator` implicitly calls `from_non_deterministic_state` if no global generator is set, and so will also raise an error if a global generator is not set with `tf.random.set_global_generator`.
+
+Stateless random functions, such as `tf.random.stateless_normal`, are always deterministic and so will never raise determinism-related errors.
+
+In graph mode, ops will raise an error message when the random op is created. If a random op is created in graph mode without determinism being enabled but then later runs when determinism is enabled, it will also raise an error.
+
+No error will be raised if a random op or generator is run before determinism is enabled (as is true for any other op), so users should take care to enable determinism before running any random ops or generators.
+
+### Parameter Server
+Use of parameter servers adds nondeterministic behavior. In case a model constructs a ParameterServerStrategy, TensorFlow will throw an error. We’ll also document this in the documentation for the flag.
+
+### Op Review and changes
+As part of the implementation, we will review all ops to make a determination of their behavior (deterministic vs nondeterministic). Some of the ops that are known to be nondeterministic, at least when running on a GPU, include:
```

Comment:
This is not a pragmatic approach and it won't materialize like this. First of all, I already have a well-bounded set of ops that have kernels that operate nondeterministically on GPU (most of the issue is on GPU and not CPU). Those ops/kernels were found by debugging nondeterminism in real models, isolating the op that was injecting the nondeterminism, and creating a self-contained repro case. For the remaining ops (including below) that I have not already enhanced to provide deterministic functionality, we just need to throw the error.

There are other GPU ops that we have confirmed use CUDA atomicAdd that are not in the list below. We can focus on those to confirm that they operate nondeterministically before having them throw an error.

There will most likely be ops we miss initially, and those will show up sometimes and introduce nondeterminism in models training with `tf.config.enable_deterministic_execution(True)`. That would be a bug in the functionality, which would then be fixed by isolating the source, confirming it, and putting putting in the error-throwing code.

I think that line 68 should be changed to:

**Ops that are known to operate nondeterministically, at least when running on a GPU, include the following:**

then, after the list of ops, I think it should read:

**We have a list of other ops that use CUDA atomicAdd and are therefore likely to be sources of nondeterminism. Once it has been confirmed that those ops operate nondeterministically, they will be made to throw errors appropriately after calling `tf.config.enable_deterministic_execution(True)`.**